src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs

```diff
@@ -903,43 +903,55 @@ protected void CheckAssigned(Symbol symbol, SyntaxNode node)
                     if (slot >= this.State.Assigned.Capacity) Normalize(ref this.State);
                     if (slot > 0 && !this.State.IsAssigned(slot))
                     {
-                        ReportUnassigned(symbol, node);
+                        ReportUnassignedIfOutsideLocalFunction(symbol, node, slot);
                     }
                 }
             }
         }
 
-        /// <param name="symbol">Symbol to variable that is unassigned.</param>
-        /// <param name="node">Syntax where read occurs.</param>
-        /// <param name="slotOpt">Optional slot where variable is located.</param>
-        /// <param name="skipIfUseBeforeDeclaration">
-        /// True if error reporting should consider the location where the
-        /// variable is declared (for instance, eliding errors about reading
-        /// variables that have not yet been declared).
-        /// </param>
-        private void ReportUnassigned(Symbol symbol, SyntaxNode node, int? slotOpt, bool skipIfUseBeforeDeclaration = true)
+        private void ReportUnassignedIfOutsideLocalFunction(Symbol symbol, SyntaxNode node, int slot, bool skipIfUseBeforeDeclaration = true)
         {
-            int slot = slotOpt ?? VariableSlot(symbol);
-            if (slot <= 0) return;
+            // If the symbol is captured by the nearest
+            // local function, record the read and skip the diagnostic
+            if ((object)GetNearestLocalFunctionOpt(currentMethodOrLambda) != null)
+            {
+                // If this is a constant, constants are always definitely assigned
+                // so we should skip reporting. This can happen in a local function
+                // where we use a constant before we actually visit its definition
+                // (since local function declarations are visited before other statements).
+                if (symbol is LocalSymbol local && local.IsConst)
```

Comment:
The comment is meant to describe this, but basically this can only happen in local functions. Otherwise, if you were to use a constant before visiting it you would get a use-before-declaration error and the definite assignment issue wouldn't be reported.